src/core/lib/gpr/tls.h

```diff
@@ -21,52 +21,143 @@
+                "unsupported type for TLS");
 
-   Accessing a thread local:
-     current_value = gpr_tls_get(&foo);
+ public:
+  using Type = T;
+};
 
-   ALL functions here may be implemented as macros. */
+}  // namespace grpc_core
 
 #ifdef GPR_STDCPP_TLS
-#include "src/core/lib/gpr/tls_stdcpp.h"
+#define GPR_THREAD_LOCAL(type) \
+  thread_local grpc_core::TlsTypeConstrainer<type>::Type
 #endif
 
 #ifdef GPR_GCC_TLS
-#include "src/core/lib/gpr/tls_gcc.h"
+#define GPR_THREAD_LOCAL(type) \
+  __thread grpc_core::TlsTypeConstrainer<type>::Type
 #endif
 
 #ifdef GPR_MSVC_TLS
-#include "src/core/lib/gpr/tls_msvc.h"
+#define GPR_THREAD_LOCAL(type) \
+  __declspec(thread) grpc_core::TlsTypeConstrainer<type>::Type
 #endif
 
 #ifdef GPR_PTHREAD_TLS
-#include "src/core/lib/gpr/tls_pthread.h"
+#include <grpc/support/log.h> /* for GPR_ASSERT */
+#include <pthread.h>
+
+namespace grpc_core {
+
+template <typename T>
+class PthreadTlsImpl : TlsTypeConstrainer<T> {
+ public:
+  PthreadTlsImpl() = default;
+  PthreadTlsImpl(const PthreadTlsImpl&) = delete;
+  PthreadTlsImpl& operator=(const PthreadTlsImpl&) = delete;
+
+  void Init() { GPR_ASSERT(0 == pthread_key_create(&key_, nullptr)); }
+
+  void Destroy() { GPR_ASSERT(0 == pthread_key_delete(key_)); }
+
+  operator T() const { return Cast<T>(pthread_getspecific(key_)); }
+
+  T operator=(T t) {
+    GPR_ASSERT(0 == pthread_setspecific(key_, Cast<T>(t)));
+    return t;
+  }
+
+ private:
+  // TODO(C++17): Replace these helpers with constexpr if statements inline.
+  template <typename V>
+  static typename std::enable_if<std::is_pointer<V>::value, V>::type Cast(
+      void* object) {
+    return static_cast<V>(object);
+  }
+
+  template <typename V>
+  static typename std::enable_if<
+      !std::is_pointer<V>::value && std::is_integral<V>::value, V>::type
+  Cast(void* object) {
+    return reinterpret_cast<uintptr_t>(object);
+  }
+
+  template <typename V>
+  static void* Cast(
+      typename std::enable_if<std::is_pointer<V>::value, V>::type t) {
+    return t;
+  }
+
+  template <typename V>
+  static void* Cast(typename std::enable_if<!std::is_pointer<V>::value &&
+                                                std::is_integral<V>::value,
+                                            V>::type t) {
+    return reinterpret_cast<void*>(uintptr_t(t));
+  }
+
+  pthread_key_t key_;
+};
+
+// This class is never instantiated. It exists to statically ensure that all
+// TLS usage is compatible with the most restrictive implementation, allowing
+// developers to write correct code regardless of the platform they develop on.
+template <typename T>
+class TriviallyDestructibleAsserter {
+  // This type is often used as a global; since the type itself is hidden by the
+  // macros, enforce compliance with the style guide here rather than at the
+  // caller. See
+  // https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables.
+  static_assert(std::is_trivially_destructible<T>::value,
```

Comment:
Can this assertion be moved into the `PthreadTlsImpl` ctor, so that we don't need a separate class here?

Reply:
The thing which must be trivially destructible is `PthreadtlsImpl<T>`, not `T` (though `T` *also* must be trivially destructible, but that's enforced separately by `TlsTypeConstrainer` and for a different reason)